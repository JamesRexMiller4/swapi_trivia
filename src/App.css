/* Colors:
  dark-grey: 131319
  starwars-yellow: EEDB00
  light-grey: EFF6EE
  light-blue: 2274A5
  dark-blue: 001F54 */


* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.App {
  height: 100vh;
  text-align: center;
  width: 100vw;
}
