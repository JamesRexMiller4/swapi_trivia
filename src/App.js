import React, { Component } from 'react';
import { Route, NavLink, Redirect } from 'react-router-dom';
import './App.css';
import Container from './components/Container/Container.js';
import Header from './components/Header/Header.js';
import Landing from './components/Landing/Landing.js';
import Loader from './components/Loader/Loader.js';


class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      path: '/',
      user: {
        name: '',
        quote: '',
        ranking: ''
      },
      movies: [],
      movieID: null,
      characters: []
    }
  }

  componentDidMount() {
    this.state.user &&
    fetch('https://swapi.co/api/films/')
      .then(res => res.json())
      .then(data => {
        let movies = data.results.map((movie, index) => {
          let arr = movie.release_date.split('-')
          let date = arr[0]
            return ({
              key: index,
              id: movie.episode_id,
              title: movie.title.toUpperCase(),
              date: date,
              characters: movie.characters,
              openingCrawl: movie.opening_crawl,
              type: 'movie'
            })
          })
          console.log(movies)
        return movies.sort((a,b) => a.id - b.id)
      })
      .then(movies => this.setState({isLoading: false, movies: movies}))
      .catch(error => console.log(error))
  }

  componentDidUpdate(prevState) {
    if (this.state.characters.length < 1 && this.state.isLoading && this.state !== prevState) {
      const movie = this.state.movies.find(movie => movie.id === parseInt(this.state.movieID))
      const fetchChars = movie.characters.map(character => {
        return fetch(character)
          .then(res => res.json())
          .then(char => {
            const charHomeData = fetch(char.homeworld)
              .then(res => res.json())
              .then(homeworld => {
                console.log('Resolved')
                return {name: char.name, homeworld: homeworld.name, population: homeworld.population}
              })
              .catch(error => console.log('homeworld fetch failed'))

            const species = fetch(char.species)
              .then(res => res.json())
              .then(species => {
                return {species: species.name}
              })
              .catch(error => console.log('species fetch failed'))

            const charMovies = char.films.map(film => {
              return fetch(film)
              .then(res => res.json())
              .then(film => film.title)
              .catch(error => 'films failed')
            })

            return Promise.all([charHomeData, species, Promise.all(charMovies)])
          })
          })
      Promise.all(fetchChars)
        .then(characters => this.setState({characters}))
        .catch(error => console.log("Something has gone wrong"))
    }
  }

  updateLogin = ({ name, quote, ranking }) => {
    this.setState({ user: { name, quote, ranking },
    path: '/movies'});
  }

  logout = () => {
    this.setState({path: '/'});
  }

  setPath = (e) => {
    let url = '/movies/' + e.target.id
    this.setState({path: url, isLoading: true, movieID: e.target.id})
  }



  render = () => {
    let moviePage;
    let characterPage;
    this.state.isLoading ? moviePage = <Loader /> :
    moviePage = <>
      <Header {...this.state.user} logout={this.logout} />
      <Container data={this.state.movies} setPath={this.setPath}/>
     </>
     !this.state.characters.length ? characterPage = <Loader /> :
     characterPage = <>
       <Header {...this.state.user} logout={this.logout} />
       <Container data={this.state.characters} setPath={this.setPath}/>
      </>
    return (
      <div className="App">
        <Redirect to={this.state.path} />
        <Route exact path='/' render={() => <Landing updateLogin={this.updateLogin} />} />
        <Route exact path='/movies' render={() => moviePage} />
        <Route exact path='/movies/:movie_id' render={() => characterPage} />
        <Route path='/favorite' render={() => <Container />} />
      </div>
    );
  }
}

export default App;
