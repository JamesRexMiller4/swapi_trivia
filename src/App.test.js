import React from 'react';
import App from './App';
import {shallow} from 'enzyme';

describe('App', () => {
  let wrapper, defaultState, formState;

  beforeEach(() => {
    defaultState = {
      isLoading: true,
      path: '/',
      user: {
        name: '',
        quote: '',
        ranking: ''
      },
      movies: [],
      movieID: null,
      characters: [],
      favoriteChars: []
    };

    formState = {
      name: 'Rick Sanchez', 
      quote: 'Wubbalubbadubdub', 
      ranking: 'expert'
    };

    wrapper = shallow(<App />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state upon mounting', () => {
    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should update state when updateLogin is called', () => {
    wrapper.instance().componentDidMount = jest.fn();
    wrapper.instance().componentDidUpdate = jest.fn();
    
    expect(wrapper.state()).toEqual(defaultState);
    wrapper.instance().setState(formState);
    wrapper.instance().updateLogin(wrapper.state());
    expect(wrapper.state('user')).toEqual(formState);
  });

  it('should update the path in state when logout is called', () => {
    wrapper.instance().componentDidMount = jest.fn();
    wrapper.instance().componentDidUpdate = jest.fn();

    wrapper.instance().setState(formState);
    wrapper.instance().updateLogin(wrapper.state());
    expect(wrapper.state('path')).toEqual('/movies');
    wrapper.instance().logout();
    expect(wrapper.state('path')).toEqual('/');
  });

  it('should update the path in state when setPath is called', () => {
    const mockEvent = {target: {id: 4}};
    wrapper.instance().componentDidMount = jest.fn();
    wrapper.instance().componentDidUpdate = jest.fn();

    wrapper.instance().setState(formState);
    wrapper.instance().updateLogin(wrapper.state());
    expect(wrapper.state('path')).toEqual('/movies');
    wrapper.instance().setPath(mockEvent);
    expect(wrapper.state('path')).toEqual('/movies/4')
  })

  it('should add a favorite character to state.favoriteChars when addFavorite is called', () => {
    const filmsArr = ['A NEW HOPE', 'RETURN OF THE JEDI', 'THE FORCE AWAKENS'];
      const favChars = [
      [{name: 'Luke', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr],
      [{name: 'Vadar', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr],
      [{name: 'C3PO', homeworld: 'Tatooine', population: '200000'}, {species: 'droid'}, filmsArr]
      ];
    const hanSOLO = [{name: 'Han', homeworld: 'Dagoba', population: '4000000'}, {species: 'human'}, filmsArr]
    const mockEvent = { preventDefault: jest.fn(), target: {id: 'Han', setAttribute: jest.fn()} };
    const expected = [favChars, hanSOLO];
  
    wrapper.instance().setState({characters: [favChars, hanSOLO]})
    wrapper.instance().setState({favoriteChars: [favChars]});
    wrapper.instance().addFavorite(mockEvent);

    expect(wrapper.state('favoriteChars')).toEqual(expected)
  });

  it('should remove a favorite character from state.favoriteChars when removeFavorite is called', () => {
    wrapper.instance().componentDidMount = jest.fn();
    wrapper.instance().componentDidUpdate = jest.fn();

    const filmsArr = ['A NEW HOPE', 'RETURN OF THE JEDI', 'THE FORCE AWAKENS'];
    const favChars = [
      [{name: 'Luke', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr],
      [{name: 'Vadar', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr],
      [{name: 'C3PO', homeworld: 'Tatooine', population: '200000'}, {species: 'droid'}, filmsArr]
      ];
    const mockEvent = { preventDefault: jest.fn(), target: {id: 'C3PO', setAttribute: jest.fn()} };
    const expected = [[{name: 'Luke', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr],
      [{name: 'Vadar', homeworld: 'Tatooine', population: '200000'}, {species: 'human'}, filmsArr]];
  
    wrapper.instance().setState({favoriteChars: favChars});
    wrapper.instance().removeFavorite(mockEvent);
    expect(wrapper.state('favoriteChars')).toEqual(expected);
  })

  it('should call either addFavorite or removeFavorite when checkFavorite is called', () => {
    wrapper.instance().componentDidMount = jest.fn();
    wrapper.instance().componentDidUpdate = jest.fn();
    wrapper.instance().addFavorite = jest.fn();
    wrapper.instance().removeFavorite = jest.fn();

    const mockEvent1 = { preventDefault: jest.fn(), target: {id: 'C3PO', class: 'heart', setAttribute: jest.fn()} };

    wrapper.instance().checkFavorite(mockEvent1);
    expect(wrapper.instance().addFavorite).toHaveBeenCalled();

    const mockEvent2 = { preventDefault: jest.fn(), target: {id: 'C3PO', class: 'heart-2', setAttribute: jest.fn()} };
    wrapper.instance().checkFavorite(mockEvent2);
    expect(wrapper.instance().removeFavorite).toHaveBeenCalled();
  });

  it.skip('should test fetch data when component did mount', () => {

  });



});
