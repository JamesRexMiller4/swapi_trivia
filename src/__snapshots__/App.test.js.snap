// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Redirect
    to="/"
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies/:movie_id"
    render={[Function]}
  />
  <Route
    path="/favorites"
    render={[Function]}
  />
</div>
`;
